src/sys/windows/selector.rs

```diff
@@ -1,596 +1,541 @@
+        )?)))
+    }
 
-macro_rules! offset_of {
-    ($t:ty, $($field:ident).+) => (
-        &(*(0 as *const $t)).$($field).+ as *const _ as usize
-    )
+    fn add_socket_to_update_queue<S: WindowsSocketState>(&self, socket: &S) {
+        let sock_state = socket.get_sock_state().unwrap();
+        let mut update_queue = self.update_queue.lock().unwrap();
+        update_queue.push_back(sock_state);
+    }
 }
 
-// See sys::windows module docs for why this exists.
-//
-// The gist of it is that `Selector` assumes that all `OVERLAPPED` pointers are
-// actually inside one of these structures so it can use the `Callback` stored
-// right after it.
-//
-// We use repr(C) here to ensure that we can assume the overlapped pointer is
-// at the start of the structure so we can just do a cast.
-/// A wrapper around an internal instance over `miow::Overlapped` which is in
-/// turn a wrapper around the Windows type `OVERLAPPED`.
-///
-/// This type is required to be used for all IOCP operations on handles that are
-/// registered with an event loop. The event loop will receive notifications
-/// over `OVERLAPPED` pointers that have completed, and it will cast that
-/// pointer to a pointer to this structure and invoke the associated callback.
-#[repr(C)]
-pub struct Overlapped {
-    inner: UnsafeCell<miow::Overlapped>,
-    callback: fn(&OVERLAPPED_ENTRY),
-}
+fn interests_to_afd_flags(interests: Interests) -> u32 {
+    let mut flags = 0;
 
-impl Overlapped {
-    /// Creates a new `Overlapped` which will invoke the provided `cb` callback
-    /// whenever it's triggered.
-    ///
-    /// The returned `Overlapped` must be used as the `OVERLAPPED` passed to all
-    /// I/O operations that are registered with mio's event loop. When the I/O
-    /// operation associated with an `OVERLAPPED` pointer completes the event
-    /// loop will invoke the function pointer provided by `cb`.
-    pub fn new(cb: fn(&OVERLAPPED_ENTRY)) -> Overlapped {
-        Overlapped {
-            inner: UnsafeCell::new(miow::Overlapped::zero()),
-            callback: cb,
-        }
+    if interests.is_readable() {
+        flags |=
+            AFD_POLL_RECEIVE | AFD_POLL_RECEIVE_EXPEDITED | AFD_POLL_ACCEPT | AFD_POLL_DISCONNECT;
     }
 
-    /// Get the underlying `Overlapped` instance as a raw pointer.
-    ///
-    /// This can be useful when only a shared borrow is held and the overlapped
-    /// pointer needs to be passed down to winapi.
-    pub fn as_mut_ptr(&self) -> *mut OVERLAPPED {
-        unsafe { (*self.inner.get()).raw() }
+    if interests.is_writable() {
+        flags |= AFD_POLL_SEND;
     }
+
+    flags
 }
 
-impl fmt::Debug for Overlapped {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Overlapped").finish()
+fn get_base_socket(raw_socket: RawSocket) -> io::Result<RawSocket> {
+    let mut base_socket: RawSocket = 0;
+    let mut bytes: u32 = 0;
+    const SIO_BASE_HANDLE: u32 = 0x48000022;
```

Comment:
This is also exported by `winapi` in `winapi::um::mswsock::SIO_BASE_HANDLE`, please use that.